test/dockerized/run.sh

```diff
@@ -0,0 +1,31 @@
+#!/usr/bin/env bash
+
+cd "$(dirname "$0")"
+
+if [[ -z $(docker container ls | grep cluster_scylla) ]]; then
+    echo "Scylla container must be running to execute tests (use 'make up')."
+    exit 1
+fi
+
+if ! [[ $(docker image ls -q scylla_rust_driver_testing) ]]; then
+    docker build -t scylla_rust_driver_testing .
+fi
+
+if [ -z "$CARGO_HOME" ]; then
+    CARGO_REGISTRY=$HOME/.cargo/registry
+else
+    CARGO_REGISTRY=$CARGO_HOME/registry
+fi
+
+docker run --name "scylla-rust-driver-testing" \
+    --network scylla_rust_driver_public \
+    -v "$(pwd)/../..:/scylla_driver:Z" \
+    -v "$CARGO_REGISTRY:/usr/local/cargo/registry:z" \
```

Comment:
This might actually cause a similar problem to the mount above. I understand that this is done to save time (so that the container doesn't have to download everything every time).

Perhaps we should create a `Dockerfile` which doesn't mount anything but rather copies files inside. There is a nice trick described in this article about caching dependencies when using a Dockerfile: http://whitfin.io/speeding-up-rust-docker-builds/ .

Reply:
Won't this method require building a new image for each `make dockerized-test` invocation?

I agree that mounting is a problem, but polluting local image store with one-time test images can be a problem too:(

Reply:
Also, I don't see an easy way to translate this http://whitfin.io/speeding-up-rust-docker-builds/ trick to our multi-workspace cargo project.

Reply:
> Won't this method require building a new image for each `make dockerized-test` invocation?
> 
> I agree that mounting is a problem, but polluting local image store with one-time test images can be a problem too:(

Maybe there is a way to cache only some of the steps, but not the last one which only builds and runs the tests?

> Also, I don't see any easy way to translate this http://whitfin.io/speeding-up-rust-docker-builds/ trick to our multi-workspace cargo project.

Won't copying the `Cargo.lock` and all `Cargo.toml`, while preserving the directory structure, work?

I was thinking about something like this:

```Dockerfile
# ...

RUN mkdir ./scylla/ ./scylla-cql/ ./scylla-proxy/
COPY ./Cargo.lock ./Cargo.lock
COPY ./Cargo.toml ./Cargo.toml
COPY ./scylla/Cargo.toml ./scylla/Cargo.coml
COPY ./scylla-cql/Cargo.toml ./scylla-cql/Cargo.toml
COPY ./scylla-proxy/Cargo.toml ./scylla-proxy/Cargo.toml

# Build the dependencies
RUN cargo build

# ...
```

Maybe it is possible to do in a smarter way with some globs so that we don't have to update the Dockerfile every time we add a new crate.

Reply:
The snippet above fails due to missing `examples` workspace member.

This is a minimal example that I came up with for `RUN cargo build` to pass.
```dockerfile
WORKDIR /driver
RUN mkdir -p ./scylla/src ./scylla-cql/src ./scylla-proxy/src ./scylla-macros/src ./examples/src
COPY ./Cargo.lock ./Cargo.lock
COPY ./Cargo.toml ./Cargo.toml

COPY ./scylla/Cargo.toml ./scylla/Cargo.toml
COPY ./scylla-cql/Cargo.toml ./scylla-cql/Cargo.toml
COPY ./scylla-proxy/Cargo.toml ./scylla-proxy/Cargo.toml
COPY ./scylla-macros/Cargo.toml ./scylla-macros/Cargo.toml
COPY ./examples/Cargo.toml ./examples/Cargo.toml

RUN touch scylla/src/lib.rs
RUN touch scylla-cql/src/lib.rs
RUN touch scylla-proxy/src/lib.rs
RUN touch scylla-macros/src/lib.rs
RUN touch examples/src/lib.rs

COPY ./scylla/benches/benchmark.rs ./scylla/benches/benchmark.rs
COPY ./scylla-cql/benches/benchmark.rs ./scylla-cql/benches/benchmark.rs

RUN cargo build

COPY  . .
```

Despite the complexity, it still does not work. Running `cargo test` in the built container fails due to mysterious missing imports, however `cargo clean && cargo test` works as it should.

I think that going further with this solution will increase the maintenance costs significantly - adding new crates would make us update this `Dockerfile`, which we don't test in CI.

Reply:
Also, the demand for dockerized tests is not known. If we find a lot of people dissatisfied with `make dockerized-test` speed, we can improve it then.
As for now, I'm in favor of simple solutions that don't add a lot of maintenance costs (just like `Dockerfile` from 4th version;)) .

Reply:
I agree, the current solution (as in the current state of the PR) seems fine and we can refine it later if somebody needs it.